crates/matrix-sdk/src/room/calls.rs

```diff
@@ -0,0 +1,115 @@
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent,
+    },
+    EventId, OwnedUserId, UserId,
+};
+use thiserror::Error;
+use tokio::sync::broadcast;
+use tracing::instrument;
+
+use crate::{event_handler::EventHandlerDropGuard, room::EventSource, Room};
+
+/// An error occurring while interacting with a call/rtc event.
+#[derive(Debug, Error)]
+pub enum CallError {
+    /// We couldn't fetch the remote notification event.
+    #[error("Couldn't fetch the remote event: {0}")]
+    Fetch(Box<crate::Error>),
+
+    /// We tried to decline an event which is not of type m.rtc.notification.
+    #[error("You cannot decline this event type.")]
+    BadEventType,
+
+    /// We tried to decline a call started by ourselves.
+    #[error("You cannot decline your own call.")]
+    DeclineOwnCall,
+
+    /// We couldn't properly deserialize the target event.
+    #[error(transparent)]
+    Deserialize(#[from] serde_json::Error),
+}
+
+impl Room {
+    /// Create a new decline call event for the target notification event id .
+    ///
+    /// The event can then be sent with [`Room::send`] or a
+    /// [`crate::send_queue::RoomSendQueue`].
+    #[instrument(skip(self), fields(room = %self.room_id()))]
+    pub async fn make_decline_call_event(
+        &self,
+        notification_event_id: &EventId,
+    ) -> Result<RtcDeclineEventContent, CallError> {
+        make_call_decline_event(self, self.own_user_id(), notification_event_id).await
+    }
+
+    /// Subscribe to decline call evvent for this room.
+    ///
+    /// The returned receiver will receive the sender UserID for each decline
+    /// for the matching notify event.
+    pub fn subscribe_to_call_decline_events(
+        &self,
+        notification_event_id: &EventId,
+    ) -> (EventHandlerDropGuard, broadcast::Receiver<OwnedUserId>) {
+        let (sender, receiver) = broadcast::channel(16);
+
+        let decline_call_event_handler_handle =
+            self.client.add_room_event_handler(self.room_id(), {
+                let own_notification_event_id = notification_event_id.to_owned();
+                move |event: SyncRtcDeclineEvent| async move {
+                    // Ignore decline for other unrelated notification events.
+                    if let Some(declined_event_id) =
+                        event.as_original().map(|ev| ev.content.relates_to.event_id.clone())
+                    {
+                        if declined_event_id == own_notification_event_id {
+                            let _ = sender.send(event.sender().to_owned());
+                        }
+                    }
+                }
+            });
+        let drop_guard = self.client().event_handler_drop_guard(decline_call_event_handler_handle);
+        (drop_guard, receiver)
+    }
+}
+
+async fn make_call_decline_event<S: EventSource>(
+    source: S,
+    own_user_id: &UserId,
+    notification_event_id: &EventId,
+) -> Result<RtcDeclineEventContent, CallError> {
```

Comment:
Is it necessary for this to be a generic function? As far as I can see you're always mocking things to get a room and never use any other `EventSource`.

Reply:
I tried to be consistent with the way `make_edit_event` is doing it, as it is also a relation.
I am not testing it to simplify the mocking by using the event cache, but I can add a test without event_cache and then mock the `/message` API?

Reply:
Well ok, but if we don't use the generics then it's a bit pointless to introduce them. They don't hurt that much but they do slow down compile times a bit.

> I am not testing it to simplify the mocking by using the event cache, but I can add a test without event_cache and then mock the /message API?

I don't think that's necessary, writing the test the way you did is a more realistic scenario. Those generics are there if it's very hard to inject the necessary test data into various places.

Reply:
simplified to be just Room